### ISSUE TITLE
ZFS auto-detect features not working 100% on FreeBSD

### ISSUE THREAD

**Issue by marcomayer** [bolthole/zrep#66]
I know I should make this a pull request but I'm short on time and have no git setup with me at the moment:

On FreeBSD some of the auto-detect features don't work properly, here's what zfs looks like on FreeBSD 11.1:

```
missing command
usage: zfs command args ...
where 'command' is one of the following:

	create [-pu] [-o property=value] ... <filesystem>
	create [-ps] [-b blocksize] [-o property=value] ... -V <size> <volume>
	destroy [-fnpRrv] <filesystem|volume>
	destroy [-dnpRrv] <filesystem|volume>@<snap>[%<snap>][,...]
	destroy <filesystem|volume>#<bookmark>

	snapshot|snap [-r] [-o property=value] ... <filesystem|volume>@<snap> ...
	rollback [-rRf] <snapshot>
	clone [-p] [-o property=value] ... <snapshot> <filesystem|volume>
	promote <clone-filesystem>
	rename [-f] <filesystem|volume|snapshot> <filesystem|volume|snapshot>
	rename [-f] -p <filesystem|volume> <filesystem|volume>
	rename -r <snapshot> <snapshot>
	rename -u [-p] <filesystem> <filesystem>	bookmark <snapshot> <bookmark>

	list [-Hp] [-r|-d max] [-o property[,...]] [-s property]...
	    [-S property]... [-t type[,...]] [filesystem|volume|snapshot] ...

	set <property=value> ... <filesystem|volume|snapshot> ...
	get [-rHp] [-d max] [-o "all" | field[,...]]
	    [-t type[,...]] [-s source[,...]]
	    <"all" | property[,...]> [filesystem|volume|snapshot] ...
	inherit [-rS] <property> <filesystem|volume|snapshot> ...
	upgrade [-v]
	upgrade [-r] [-V version] <-a | filesystem ...>
	userspace [-Hinp] [-o field[,...]] [-s field] ...
	    [-S field] ... [-t type[,...]] <filesystem|snapshot>
	groupspace [-Hinp] [-o field[,...]] [-s field] ...
	    [-S field] ... [-t type[,...]] <filesystem|snapshot>

	mount
	mount [-vO] [-o opts] <-a | filesystem>
	unmount|umount [-f] <-a | filesystem|mountpoint>
	share <-a | filesystem>
	unshare <-a | filesystem|mountpoint>

	send [-DnPpRvLe] [-[iI] snapshot] <snapshot>
	send [-Le] [-i snapshot|bookmark] <filesystem|volume|snapshot>
	send [-nvPe] -t <receive_resume_token>
	receive|recv [-vnsFu] <filesystem|volume|snapshot>
	receive|recv [-vnsFu] [-o origin=<snapshot>] [-d | -e] <filesystem>
	receive|recv -A <filesystem|volume>

	allow <filesystem|volume>
	allow [-ldug] <"everyone"|user|group>[,...] <perm|@setname>[,...]
	    <filesystem|volume>
	allow [-ld] -e <perm|@setname>[,...] <filesystem|volume>
	allow -c <perm|@setname>[,...] <filesystem|volume>
	allow -s @setname <perm|@setname>[,...] <filesystem|volume>

	unallow [-rldug] <"everyone"|user|group>[,...]
	    [<perm|@setname>[,...]] <filesystem|volume>
	unallow [-rld] -e [<perm|@setname>[,...]] <filesystem|volume>
	unallow [-r] -c [<perm|@setname>[,...]] <filesystem|volume>
	unallow [-r] -s @setname [<perm|@setname>[,...]] <filesystem|volume>

	hold [-r] <tag> <snapshot> ...
	holds [-Hp] [-r|-d depth] <filesystem|volume|snapshot> ...
	release [-r] <tag> <snapshot> ...
	diff [-FHt] <snapshot> [snapshot|filesystem]

	jail <jailid|jailname> <filesystem>
	unjail <jailid|jailname> <filesystem>

Each dataset is of the form: pool/[dataset/]*dataset[@name]

For the property list, run: zfs set|get

For the delegated permission list, run: zfs allow|unallow

```

The following fixes the ones I found to be wrong:
```

201c201
< 	if grep 'receive .*-[a-zA-Z]*x' $zrep_checkfile >/dev/null ;then
---
> 	if grep 'receive.*-[a-zA-Z]*x' $zrep_checkfile >/dev/null ;then
204c204
< 	if grep 'receive .*-[a-zA-Z]*u' $zrep_checkfile >/dev/null ;then
---
> 	if grep 'receive.*-[a-zA-Z]*u' $zrep_checkfile >/dev/null ;then
```
The next one is a difficult one. zfs create has -o on FreeBSD but using -o with receive has another meaning:

> 
> -o origin=snapshot
>                  Forces the stream to be received as a clone of the given
>                  snapshot.  If the stream is a full send stream, this will
>                  create the filesystem described by the stream as a clone of
>                  the specified snapshot. Which snapshot was specified will not
>                  affect the success or failure of the receive, as long as the
>                  snapshot does exist.  If the stream is an incremental send
>                  stream, all the normal verification will be performed.

So it actually won't work when zrep uses "receive -o":
```
208c208
< 	if grep 'create .* -o prop' $zrep_checkfile >/dev/null ;then
---
> 	if grep 'create .*-o prop' $zrep_checkfile >/dev/null ;then
```

Maybe there should be two variables for this, Z_HAS_CREATE_O and Z_HAS_REC_O to distinguish.

**Comment by ppbrown** [comment]
sorry for lag on reply.

I updated Z_HAS_X check.
commit 992e6b3

the -o bit, is tricky.
Lets see if I understand what you are saying.
Are you saying that zrep incorrectly thinks that freebsd CAN use
receive -o readonly=on
when in actuality, it cannot.... so I need to somehow turn OFF use?

(maybe it's like SmartOS? Hmm, if only i had the zfs usage output from that to compare...)

**Comment by ppbrown** [comment]
actually.. i DO have it.

issue #35

**Comment by ppbrown** [comment]
patched in 77d5356

**Comment by ppbrown** [comment]
huh. and turns out I have an archive, in the src tree, under "usage".
added freebsd. thanks!
